lib_eio/pool.ml

```diff
@@ -0,0 +1,154 @@
+type 'a check = 'a -> bool
+
+type 'a dispose = 'a -> unit
+
+type 'a handlers = {
+  check : 'a check;
+  dispose : 'a dispose;
+}
+
+let noop_handlers : 'a handlers = {
+  check = (fun _ -> true);
+  dispose = (fun _ -> ());
+}
+
+type 'a ready = 'a * 'a handlers
+
+type 'a alloc = unit -> 'a ready
+
+type 'a t = {
+  max_size : int;
+  alloc_budget : Semaphore.t;
+  alloc : 'a alloc;
+  waiting : 'a ready Promise.u option Stream.t;
+  ready : 'a ready Stream.t;
+  (* Each runner is given a copy of the clear signal
+     to be checked after its run
+
+     The copy to use is replaced after each invocation
+     of [clear], so we can distinguish between
+     clearing of different batch of runs, e.g.
+
+     1. use (clear signal version 0)
+     2. use (clear signal version 0)
+     3. clear
+     4. use (clear signal version 1)
+
+     clear at 3 should apply to use at 1 and 2, but not use at 4
+  *)
+  mutable clear_signal : bool Atomic.t;
+  shutdown : bool Atomic.t;
+}
+
+let start_monitor ~sw (t : 'a t) : unit =
+  let rec aux () =
+    match Stream.take t.waiting with
```

Comment:
With this current API, a pool must be explicitly `shutdown` in order for the forked monitoring fiber to finish and for the program to move past whatever switch is used to create the pool. I hit this problem right away when using the pool:

```ocaml
open Eio

let () =
  Eio_main.run @@ fun _ ->
  let test_pool () =
    Eio.Switch.run @@ fun sw ->
    let p = Pool.create ~sw ~alloc:(fun () -> (), Pool.noop_handlers) 2 in
    let use () =
      Eio.Fiber.fork ~sw @@ fun () ->
      Pool.use p (fun () -> Eio.traceln "Run!"; Eio_unix.sleep 2.)
    in
    let runs = [ (); (); (); (); (); () ] in
    List.iter use runs;
    (* Pool.shutdown p *)
  in
  test_pool ();
  Eio.traceln "All done"
```

This program hangs without explicitly shutting down the pool. I think this might surprise some people, is there a way to avoid this?

Reply:
This actually caught myself off guard as well, so I agree it is not very good design.

I think `shutdown` can be replaced by polling from Eio switch? I'll have to look at other Eio modules.

Reply:
I wonder if the pool itself needs to spawn the monitoring fiber, or if it just needs to return a resource when a user calls `use` which would yield when the pool is out of resources. This would avoid needing to pass in this switch.

Reply:
I think my original intention was to make analysis simple, especially in anticipation of faults in allocation, release etc. I don't know how faithful I've stuck to that intention, and it could be the case that we can leave the responsibilities to `use`.

I'm leaning toward having a monitor for simpler code (in my attempts thus far), but would not mind swapping around approaches.

Reply:
You might be able to keep most things the same and just use `Fiber.fork_daemon` instead and this wouldn't (I think) be an issue anymore https://ocaml.org/p/eio/latest/doc/Eio/Fiber/index.html#val-fork_daemon

Reply:
Ah right, it carries the exact behaviour we want for monitor. Thanks!